sharktank/sharktank/tools/tuner/candidate_gen.py

```diff
@@ -0,0 +1,1312 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Given an input dispatch, this code modifies the hyperparameters
+# in the code and runs it.
+
+import argparse
+import logging
+import math
+import pickle
+import re
+import z3
+from dataclasses import asdict, dataclass
+from enum import Enum
+from os import mkdir, path, makedirs
+from typing import Callable
+from textwrap import indent
+
+import iree.compiler as ireec
+from iree.compiler import ir
+from iree.compiler.dialects import _linalg_ops_gen, _util_ops_gen
+
+"""
+Usage: ./candidate_gen.py 121.mlir -o "tuning/candidates" -l 1024 --lhs-dims=mk --rhs-dims=nk --tile-dims=mnk
+"""
```

Comment:
I'd merge this documentation with the comment above, rather than separate the two with all the `import` statements.
```python
# Given an input dispatch, this code modifies the hyperparameters
# in the code and runs it.

import foo  # <-- these are right between the docs
import bar

"""
Usage: ./candidate_gen.py 121.mlir -o "tuning/candidates" -l 1024 --lhs-dims=mk --rhs-dims=nk --tile-dims=mnk
"""
```

Maybe following this format, all above the `import`s:
```python
"""(One line description of what purpose the script has)

(Other context, information about prerequisites, links to documentation, etc.)
(More context...)

Usage: ./candidate_gen.py 121.mlir -o "tuning/candidates" -l 1024 --lhs-dims=mk --rhs-dims=nk --tile-dims=mnk
"""
```